alf/algorithms/generator_test.py

```diff
@@ -0,0 +1,180 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import unittest
+
+from absl import logging
+from absl.testing import parameterized
+import tensorflow as tf
+
+from tf_agents.networks.network import Network
+from alf.algorithms.generator import Generator
+from alf.algorithms.mi_estimator import MIEstimator
+
+
+class Net(Network):
+    def __init__(self, dim):
+        super().__init__(
+            input_tensor_spec=tf.TensorSpec(shape=(dim, )),
+            state_spec=(),
+            name="Net")
+        self._w = tf.Variable(
+            initial_value=[[1, 2], [1, 1]], shape=(dim, dim), dtype=tf.float32)
+
+    def call(self, input):
+        return tf.matmul(input, self._w), ()
+
+
+class Net2(Network):
+    def __init__(self, dim):
+        super().__init__(
+            input_tensor_spec=[
+                tf.TensorSpec(shape=(dim, )),
+                tf.TensorSpec(shape=(dim, ))
+            ],
+            state_spec=(),
+            name="Net")
+        self._w = tf.Variable(
+            initial_value=[[1, 2], [1, 1]], shape=(dim, dim), dtype=tf.float32)
+        self._u = tf.Variable(
+            initial_value=tf.zeros((dim, dim)),
+            shape=(dim, dim),
+            dtype=tf.float32)
+
+    def call(self, input):
+        return tf.matmul(input[0], self._w) + tf.matmul(input[1], self._u), ()
+
+
+class GeneratorTest(parameterized.TestCase, unittest.TestCase):
+    def assertArrayEqual(self, x, y, eps):
+        self.assertEqual(x.shape, y.shape)
+        self.assertLessEqual(float(tf.reduce_max(abs(x - y))), eps)
+
+    @parameterized.parameters(
+        dict(mode='STEIN'),
+        dict(mode='ML'),
+        dict(mode='ML', mi_weight=1),
+    )
+    def test_generator_unconditional(self, mode='ML', mi_weight=None):
+        """
+        The generator is trained to match(STEIN)/maximize(ML) the likelihood
+        of a Gaussian distribution with zero mean and diagonal variance (1, 4).
+        After training, w^T w is the variance of the distribution implied by the
+        generator. So it should be diag(1,4) for STEIN and 0 for 'ML'.
+        """
+        logging.info("mode: %s mi_weight: %s" % (mode, mi_weight))
+        dim = 2
+        batch_size = 512
+        net = Net(dim)
+        generator = Generator(
+            dim,
+            noise_dim=dim,
```

Comment:
add test case where out_dim != noise_dim to check the bug

Reply:
done